odk/collect/collect_app/src/main/java/org/odk/collect/android/openrosa/okhttp/OkHttpConnection.java

```diff
@@ -206,36 +210,56 @@ private HttpPostResult executePostRequest(@NonNull URI uri, @Nullable HttpCreden
 
         Response response;
         HttpPostResult postResult;
+        HttpPostResult failedResult = new HttpPostResult(
+                "Form Upload Failed",
+                HttpURLConnection.HTTP_BAD_REQUEST,
+                "Problem with server configuration");
 
         if(generalSettings.getBoolean(ProjectKeys.KEY_CUSTOM_SERVER_IS_ENABLED)){
-
-            JSONObject headersInfo = new JSONObject(generalSettings.getString(ProjectKeys.KEY_CUSTOM_SERVER_HEADERS));
-            OkHttpClient client = new OkHttpClient();
+            String customHeaders = generalSettings.getString(ProjectKeys.KEY_CUSTOM_SERVER_HEADERS);
             Headers.Builder builder = new Headers.Builder();
-
-            Iterator<String> keys = headersInfo.keys();
-
-            // Iterate through the keys
-            while (keys.hasNext()) {
-                String key = keys.next();
-                String value = headersInfo.getString(key);
-                builder.add(key, value);
+            if (customHeaders != null && !customHeaders.trim().isEmpty()) {
+                try {
+                    JSONObject headersInfo = new JSONObject(customHeaders);
+                    Iterator<String> keys = headersInfo.keys();
+                    // Iterate through the keys
+                    while (keys.hasNext()) {
+                        String key = keys.next();
+                        String value = headersInfo.getString(key);
+                        builder.add(key, value);
+                    }
+                }
+                catch (JSONException e) {
+                    FormEventBus.INSTANCE.formUploadError(
+                            // TODO: return formId
```

Comment:
If possible lets pass, instance id in error instead of form id, as multiple instances of same form can be there